src/Base/Array4.H

```diff
@@ -0,0 +1,279 @@
+            // __array_function__ feature requires NumPy 1.16 or later.
+
+
+            // Nvidia GPUs: __cuda_array_interface__ v3
+            // https://numba.readthedocs.io/en/latest/cuda/cuda_array_interface.html
+            .def_property_readonly("__cuda_array_interface__", [](Array4<T> const & a4) {
+                auto d = pyAMReX::array_interface(a4);
+
+                // data:
+                // Because the user of the interface may or may not be in the same context, the most common case is to use cuPointerGetAttribute with CU_POINTER_ATTRIBUTE_DEVICE_POINTER in the CUDA driver API (or the equivalent CUDA Runtime API) to retrieve a device pointer that is usable in the currently active context.
+                // TODO For zero-size arrays, use 0 here.
+
+                // None or integer
+                // An optional stream upon which synchronization must take place at the point of consumption, either by synchronizing on the stream or enqueuing operations on the data on the given stream. Integer values in this entry are as follows:
+                //   0: This is disallowed as it would be ambiguous between None and the default stream, and also between the legacy and per-thread default streams. Any use case where 0 might be given should either use None, 1, or 2 instead for clarity.
+                //   1: The legacy default stream.
+                //   2: The per-thread default stream.
+                //   Any other integer: a cudaStream_t represented as a Python integer.
+                //   When None, no synchronization is required.
+                d["stream"] = py::none();
+
+                d["version"] = 3;
+                return d;
+            })
+
+
+            // TODO: __dlpack__ __dlpack_device__
+            // DLPack protocol (CPU, NVIDIA GPU, AMD GPU, Intel GPU, etc.)
+            // https://dmlc.github.io/dlpack/latest/
+            // https://data-apis.org/array-api/latest/design_topics/data_interchange.html
+            // https://github.com/data-apis/consortium-feedback/issues/1
+            // https://github.com/dmlc/dlpack/blob/master/include/dlpack/dlpack.h
+            // https://docs.cupy.dev/en/stable/user_guide/interoperability.html#dlpack-data-exchange-protocol
+
+            .def("to_host", [](Array4<T> const & a4) {
+                // py::tuple to std::vector
+                auto const a4i = pyAMReX::array_interface(a4);
+                auto shape = py::cast<std::vector<py::ssize_t>>(a4i["shape"]);
+                auto strides = py::cast<std::vector<py::ssize_t>>(a4i["strides"]);
+
+                // allocate host memory copy
+                auto h_data = py::array_t<std::remove_cv_t<T>>(shape, strides);
+
+                // sync copy: host data is unpinned
+                Gpu::copy(Gpu::deviceToHost,
+                    a4.dataPtr(), a4.dataPtr() + a4.size(),
+                    h_data.mutable_data()
+                );
+                return h_data;
```

Comment:
@WeiqunZhang does this look correct to you?

Reply:
Yes

Reply:
At the BaseFab/Array4 level, there is no concept of ghost cells in the BaseFab class.

Reply:
Ah, my strides are still in bytes and need to be in elements for `py::array_t`.

Reply:
I see. But is shape in elements? If so, that seems inconsistent.

Reply:
Oops, no... I forgot to set my desired value in all components of the multi fab